doc/architecture/005-Message-architecture.md

```diff
@@ -0,0 +1,160 @@
+
+      # This is not the final user interface, but the return value of something like
+      # nixopsProjects.default = nixops4.lib.mkProject { } ./deployment.nix;
+      nixopsProjects.default = { __mkResource, resources, ... }: {
+        
+        resources = {
+          # __mkResource would not be called by the user, but by an expression in `inputs.nixops4-aws.lib`, which passes the `properties` along.
+          bucket = __mkResource {
+            provider = inputs.nixops4-aws.packages.${system}.nixops4-provider;
+            args = ["s3-bucket"];
+            properties = {
+              auth = {
+                # ...
+              };
+              name = "my-bucket";
+              region = "us-west-1";
+            };
+          };
+
+          instance = __mkResource {
+            provider = inputs.nixops4-aws.packages.${system}.nixops4-provider;
+            args = ["ec2-instance"];
+            properties = {
+              auth = {
+                # ...
+              };
+              region = "us-west-1";
+              instanceType = "t3.micro";
+              # ...
+            };
+          };
+
+          nixos = __mkResource {
+            provider = inputs.nixops4.packages.${system}.nixops4-local-command;
+            properties = {
+              command = sshWrapper { # akin to https://docs.hercules-ci.com/hercules-ci-effects/reference/nix-functions/ssh
+                destination = "root@${resources.instance.ip}";
+                command = pkgs.writeScript "deploy-nixos" ''
+                  ${(nixpkgs.lib.nixosSystem { modules = [
+                      { services.myapp.settings.bucket = resources.bucket.url; }
+                    ]; }).config.system.toplevel}/bin/apply
+                '';
+              };
+            };
+          };
+
+          dns = __mkResource {
+            provider = inputs.nixops4.packages.${system}.nixops4-aws;
+            args = ["route53-record"];
+            properties = {
+              auth = {
+                # ...
+              };
+              zone = "example.com";
+              name = "myapp.example.com";
+              type = "A";
+              ttl = 300;
+              records = [resources.instance.ip];
+            };
+          };
+        };
+      };
+    };
+}
+```
+
+</details>
+
+## Decision
+
+When the users calls `nix run .#nixops deploy`, the following takes place.
+
+1. Nix builds NixOps4 and runs it.
+2. `nixops4` executable starts its subcommand `deploy`
+3. `deploy` loads the `nixopsProjects.default` expression
+  - Create the evaluator process
+  - Tell the evaluator process to load the `default` project, causing it to evaluate that attribute
+4. `deploy` queries the evaluator for the resources
+  - The evaluator invokes the `nixopsProjects.default` function, passing it the internally defined `__mkResource` function (a Nix "primop"), and a `resources` attribute set containing thunks that refer to unnamed primops.
```

Comment:
a `resources` attribute set containing thunks that refer to unnamed primops

This was very confusing until you explained to me what it meant.

The values of the `resources` attrset are not representable as pure Nix values. They require evaluation of rust code to produce the nix (primop) value that will be accessible later. This is more clearly explained in the step-by-step example below.